### ISSUE TITLE
how to do repeat like repeat in pytorch?

### ISSUE THREAD

**Issue by charliedream1** [NVIDIA/TensorRT#2408]
## Description

I didn't find repeat function in python API. I have a tensor A with shape of (1,3,512), and a tensor B with dynamic shape (-1,3,512),
how to repeat copy tensor A on axis 0 to have same shape with tensor B 

## Environment

**TensorRT Version**:  8.4.3
**NVIDIA GPU**:  v100
**NVIDIA Driver Version**: 
**CUDA Version**: 
**CUDNN Version**: 
**Operating System**: 
**Python Version (if applicable)**: 3.8
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by charliedream1** [comment]
thanks for reply。but how to get dynamic shape of B on axis 0，and loop N times to concat A to make A and B with same shape on axis 0？



---Original---
From: "Zero ***@***.***&gt;
Date: Sat, Oct 22, 2022 00:26 AM
To: ***@***.***&gt;;
Cc: "Optimus ***@***.******@***.***&gt;;
Subject: Re: [NVIDIA/TensorRT] how to do repeat like repeat in pytorch? (Issue#2408)




 
You can use https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_network_definition.html#a06099b5eb208df6817f576845e5f3edb

**Comment by zerollzeng** [comment]
Please refer to https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_network_definition.html
Use IShapeLayer, ISliceLayer, IConcanationLayer and Ilooplayer should be feasible.

@jhalakp-nvidia may know more about it.

**Comment by charliedream1** [comment]
if there's an example，it would be better.API document is not very comprehensive to understand.&nbsp;


Compared pytotch，it's so complicated to define a network，is there a plan to make it easier to use in the future？



---Original---
From: "Zero ***@***.***&gt;
Date: Sat, Oct 22, 2022 22:51 PM
To: ***@***.***&gt;;
Cc: "Optimus ***@***.******@***.***&gt;;
Subject: Re: [NVIDIA/TensorRT] how to do repeat like repeat in pytorch? (Issue#2408)




 
Please refer to https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_network_definition.html
 Use IShapeLayer, ISliceLayer, IConcanationLayer and Ilooplayer should be feasible.
 
@jhalakp-nvidia may know more about it.

**Comment by zerollzeng** [comment]
TensorRT is highly optimized for performance and it didn't target any other frameworks for API design, and it didn't support much operators like in pytorch due to the effort on the performance optimization.

I would suggest to export to onnx first and see what the torch.repeat will be interpreted in ONNX IR. and take our onnx-tensorrt as an example. you can refer to https://github.com/onnx/onnx-tensorrt/blob/main/builtin_op_importers.cpp

I don't think we will switch to the Pytorch-Style(or other styles) API in the short term due to various reasons like backward compatibility, internal design, and extra engineering effort...

**Comment by nvpohanh** [comment]
As Zero said, the recommended way is to use ONNX IR.

**Comment by ttyio** [comment]
Closing since no activity for more than 3 weeks, please reopen if you still have question, thanks!

**Comment by sean-wade** [comment]
I found that simply using network->addSlice().
But you must add setMode(nvinfer1::SliceMode::kWRAP) after it.

For your case, I think you can try this:

ISliceLayer* repeat_A = network->addSlice(
        A,
        Dims3(0, 0, 0),
        Dims3(X_TO_REPEAT, 3, 512),
        Dims3(1, 1, 1));
repeat_A->setMode(nvinfer1::SliceMode::kWRAP);

**Comment by charliedream1** [comment]
Thx



---Original---
From: ***@***.***&gt;
Date: Wed, Apr 19, 2023 17:54 PM
To: ***@***.***&gt;;
Cc: "Optimus ***@***.******@***.***&gt;;
Subject: Re: [NVIDIA/TensorRT] how to do repeat like repeat in pytorch? (Issue#2408)




 
I found that simply using network-&gt;addSlice().
 But you must add setMode(nvinfer1::SliceMode::kWRAP) after it.
 
For your case, I think you can try this:
 
ISliceLayer* repeat_A = network-&gt;addSlice(
 A,
 Dims3(0, 0, 0),
 Dims3(X_TO_REPEAT, 3, 512),
 Dims3(1, 1, 1));
 repeat_A-&gt;setMode(nvinfer1::SliceMode::kWRAP);
